ads/google/a4a/utils.js

```diff
@@ -357,92 +356,71 @@ export function additionalDimensions(win, viewportSize) {
  *
+ * @param {!../../../extensions/amp-a4a/0.1/amp-a4a.AmpA4A} a4a
  * @param {!../../../src/service/xhr-impl.FetchResponseHeaders} responseHeaders
  *   XHR service FetchResponseHeaders object containing the response
  *   headers.
- * @param {!../../../src/service/extensions-impl.Extensions} extensions
- * @return {?AmpAnalyticsConfigDef} config or null if invalid/missing.
+ * @return {?JSONType} config or null if invalid/missing.
  */
-export function extractAmpAnalyticsConfig(responseHeaders, extensions) {
+export function extractAmpAnalyticsConfig(a4a, responseHeaders) {
   if (responseHeaders.has(AMP_ANALYTICS_HEADER)) {
     try {
       const analyticsConfig =
         JSON.parse(responseHeaders.get(AMP_ANALYTICS_HEADER));
-      dev().assert(Array.isArray(analyticsConfig['url']));
-      if (analyticsConfig.url.length) {
-        extensions.loadExtension('amp-analytics');
+      console.log('anayticsConfig is ', analyticsConfig['url']);
+      dev().assert(Array.isArray(analyticsConfig['url']) &&
+          analyticsConfig['url'].length);
+      const urls = analyticsConfig.url;
+
+      const visibilitySpec = {
+        'selector': 'amp-ad',
+        'selectionMethod': 'closest',
+        'visiblePercentageMin': 50,
+        'continuousTimeMin': 1000,
+      };
+      const config = {
+        'transport': {'beacon': false, 'xhrpost': false},
+        'triggers': {
+          'continuousVisible': {
+            'on': 'visible',
+            visibilitySpec,
+          },
+          'continuousVisibleIniLoad': {
+            'on': 'ini-load',
+            visibilitySpec,
+          },
+          'continuousVisibleRenderStart': {
+            'on': 'render-start',
+            visibilitySpec,
+          },
+        },
+      };
+      const requests = {};
+      for (let idx = 1; idx <= urls.length; idx++) {
+        // TODO: Ensure url is valid and not freeform JS?
+        requests[`visibility${idx}`] = `${urls[idx - 1]}`;
       }
-      return {urls: analyticsConfig.url};
+      console.log('config is ', config);
+      // Security review needed here.
+      config['requests'] = requests;
+      config['triggers']['continuousVisible']['request'] =
+          Object.keys(requests);
+      // Add CSI pingbacks.
+      const correlator = getCorrelator(a4a.win);
+      const slotId = a4a.element.getAttribute('data-amp-slot-index');
+      const qqid = responseHeaders ? responseHeaders.get(QQID_HEADER) : 'null';
+      config['requests']['visibilityCsi'] =
+          'https://csi.gstatic.com/csi?fromAnalytics=1' +
+          `&c=${correlator}&slotId=${slotId}&qqid.0=${qqid}`;
+      config['triggers']['continuousVisibleIniLoad']['request'] =
+          'visibilityCsi';
+      config['triggers']['continuousVisibleRenderStart']['request'] =
+          'visibilityCsi';
+      return config;
     } catch (err) {
       dev().error('AMP-A4A', 'Invalid analytics', err,
```

Comment:
I am pretty sure we do collect dev error. However I fail to find where the logic is 😢